### PR TITLE
Securing math operations from overflows

### PR DESCRIPTION
**Pull request by kaliberpoziomka**
This PR should secure math operations from overflows.

### CODE REVIEW

**Review by achab — APPROVED** [review]
Thanks for this effort! I reviewed variable names and it looks OK!

File: contracts/protocol/libraries/math/math_utils.cairo [hunk 1]
```diff
@@ -28,11 +28,17 @@ namespace MathUtils:
         let (time_delta) = uint256_sub(
             Uint256(current_timestamp, 0), Uint256(last_update_timestamp, 0)
         )
-        let (temp_result, _) = uint256_mul(rate, time_delta)
+        let (temp_result, mul_overflow) = uint256_mul(rate, time_delta)
```
Comment (hsouf) [inline contracts/protocol/libraries/math/math_utils.cairo:31]: why not use `mul` on `SafeUint256` instead? it performs exactly the same checks.

File: contracts/protocol/libraries/math/math_utils.cairo [hunk 2]
```diff
@@ -28,11 +28,17 @@ namespace MathUtils:
...
+        with_attr error_message("Multiplication overflow"):
+            assert mul_overflow = Uint256(0, 0)
```
Comment (hsouf) [inline contracts/protocol/libraries/math/math_utils.cairo:33]: `mul_overflow`->`overflow`

File: contracts/protocol/libraries/math/math_utils.cairo [hunk 3]
```diff
@@ -28,11 +28,17 @@ namespace MathUtils:
...
+        end
 
         let (result, _) = uint256_unsigned_div_rem(temp_result, Uint256(SECONDS_PER_YEAR, 0))
 
-        let (interest, _) = uint256_add(Uint256(WadRayMath.RAY, 0), result)
+        let (interest, add_overflow) = uint256_add(Uint256(WadRayMath.RAY, 0), result)
```
Comment (hsouf) [inline contracts/protocol/libraries/math/math_utils.cairo:38]: same here

File: contracts/protocol/libraries/math/math_utils.cairo (hunk 1 above)
Comment (hsouf) [reply contracts/protocol/libraries/math/math_utils.cairo:31]: let's remove the op tag`mul_overflow`->`overflow`

File: contracts/protocol/libraries/math/math_utils.cairo [hunk 4]
```diff
@@ -82,9 +88,12 @@ namespace MathUtils:
         let (exp_minus_two) = uint256_checked_sub_return_zero_when_lt(exp, Uint256(2, 0))
 
         let (rate_square) = WadRayMath.ray_mul(rate, rate)
-        let (period_square, _) = uint256_mul(
+        let (period_square, mul_overflow_1) = uint256_mul(
```
Comment (hsouf) [inline contracts/protocol/libraries/math/math_utils.cairo:91]: using SafeUint256 will save you from the hassle of coming up with different names for overflows on the same file

**Review by hsouf — CHANGES_REQUESTED** [review]
Thank you @kaliberpoziomka for this PR. I suggest we use the SafeUint256 library or implement our own to use it on the rest of the code.

File: contracts/protocol/libraries/math/math_utils.cairo (hunk 4 above)
Comment (kaliberpoziomka) [reply contracts/protocol/libraries/math/math_utils.cairo]: Yes, I agree with all the comments

File: contracts/protocol/libraries/math/math_utils.cairo [hunk 5]
```diff
@@ -1,16 +1,12 @@
+%lang starknet
```
Comment (achab) [inline contracts/protocol/libraries/math/math_utils.cairo:1]: why is this line needed now?

File: contracts/protocol/libraries/math/percentage_math.cairo [hunk 6]
```diff
@@ -44,10 +44,16 @@ namespace PercentageMath:
             assert is_value_not_overflowing = TRUE
         end
 
-        let (value_mul_percentage, _) = uint256_mul(value, percentage)
-        let (intermediate_res_1, _) = uint256_add(
+        let (value_mul_percentage, mul_overflow) = uint256_mul(value, percentage)
```
Comment (achab) [inline contracts/protocol/libraries/math/percentage_math.cairo:47]: we should also use `SafeUint256.mul` here, right?

File: contracts/protocol/libraries/math/math_utils.cairo (hunk 5 above)
Comment (kaliberpoziomka) [reply contracts/protocol/libraries/math/math_utils.cairo:1]: when importing file with this %lang specification we need to specify it too. OZ SafeUint256 got this, so I needed to add this as well

File: contracts/protocol/libraries/math/percentage_math.cairo (hunk 6 above)
Comment (kaliberpoziomka) [reply contracts/protocol/libraries/math/percentage_math.cairo:47]: yes, thank you

File: contracts/protocol/libraries/math/math_utils.cairo (hunk 5 above)
Comment (achab) [reply contracts/protocol/libraries/math/math_utils.cairo:1]: oh ok got it! thanks

**Review by hsouf — APPROVED** [review]

File: tests/unit_tests/protocol/libraries/math/test_percentage_math.cairo [hunk 7]
```diff
@@ -1,5 +1,6 @@
 %lang starknet
 
+from starkware.cairo.common.cairo_builtins import HashBuiltin
```
Comment (achab) [inline tests/unit_tests/protocol/libraries/math/test_percentage_math.cairo:3]: why this? the test didn't pass?

**Pull request merged**